### DISCUSSION TITLE
Breaking Polymorphic

### DISCUSSION THREAD

**Discussion by BobACollins** [sqlalchemy/sqlalchemy#8400] [Usage Questions] [answered]
I'm trying to get my application finally upgraded sqlalchemy 1.4. I'm blocked on what looks like my final query. I have two joined table inheritance tables that we were joining with a union query to get a set of shared data out of them. [The stricter inheritance mapping changes](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#stricter-behavior-when-querying-inheritance-mappings-using-custom-queries) in 1.4 have broken this workflow and I'm not sure what I'm supposed to do to pull unions of polymorphic tables. Any help appreciated!

I built a minimal sample app to understand my issue it'll return the below error in 1.4.40 or succeed in 1.3.24

```python
from sqlalchemy import (
    create_engine,
    Column,
    Integer,
    Text,
    ForeignKey,
)

from sqlalchemy.orm import sessionmaker

from sqlalchemy.ext.declarative import declarative_base

engine = create_engine("sqlite:///:memory:")
Base = declarative_base()


class Car(Base):
    __tablename__ = "car"

    id = Column(Integer, primary_key=True)
    type = Column(Text)

    __mapper_args__ = {
        "polymorphic_identity": "car",
        "polymorphic_on": type,
    }


class Sedan(Car):
    __tablename__ = "sedan"

    id = Column(Integer, ForeignKey("car.id"), primary_key=True)
    wheel_count = Column(Integer)

    __mapper_args__ = {
        "polymorphic_identity": "sedan",
    }


class Truck(Car):
    __tablename__ = "truck"

    id = Column(Integer, ForeignKey("car.id"), primary_key=True)
    wheel_count = Column(Integer)

    __mapper_args__ = {
        "polymorphic_identity": "truck",
    }


Base.metadata.create_all(engine)
connection = engine.connect()
session = sessionmaker()(bind=engine)

session.add(Sedan(wheel_count=3))
session.add(Truck(wheel_count=4))

session.commit()

sedan_query = session.query(Sedan)
truck_query = session.query(Truck)

query = sedan_query.union(truck_query)
query.all()
```
Submapper error in 1.4.40:

```
sqlalchemy.exc.InvalidRequestError: Row with identity key (<class '__main__.Car'>, (2,), None) can't be loaded into an object; the polymorphic discriminator column 'anon_1.car_type' refers to mapped class Truck->truck, which is not a sub-mapper of the requested mapped class Sedan->sedan
```

**Comment by zzzeek** [comment, marked as answer]
well, I dont have a good answer for that, you are doing this query that while it looks very plain and obvious, I never imagined someone doing a query like that and we have no tests of any kind for such a thing, where the weird thing you are doing is creating a query against a certain subclass, then UNIONING a sibling class onto it, which from the POV of how the query works is not supposed to work, because Truck is not a Sedan.

If we look at your Query in 1.3, it has the wrong state, yet still works somehow:

```py
(Pdb) query = sedan_query.union(truck_query).union(limo_query)
(Pdb) query._entities[0].mapper
<Mapper at 0x7f7178aada80; Sedan>

```

see that's wrong - that's a Query that expects to load Sedan objects, that would ideally say "Car" since that's the common subtype of what you want to load, which it can't, because it wouldnt be able to guess that "load sedans and trucks == load cars".  then it sees "truck" and "limo" in the incoming rows and somehow something something it seems to work, I would be almost certain that pattern has its limits because the Query is hobbling along not knowing what it wants to do.  but whether or not, that Query isn't really "working" internally so I can't really try to restore the "working" behavior here because it doesn't make any sense.

it's hard for me to defend that because it *looks* so simple and obvious what you have there, and when I show you the way I had envisioned such a query being done, it will not look as nice.   In 1.3, I did not even imagine you could do a query like that where you just tack on a sibling class.  I was even hoping to illustrate how that form doesn't work with more variations, adding on a third class Limo with the same idea, and even that worked.    Then I tried going into 1.4 and just removing the assertion I have there..maybe it asserts some other case, but the case you have should work.  nope, doesn't work, returns an incorrect Sedan object (which is what I'd expect your query to do), and even trying to add Limo onto it, it fails even worse because it can't build the statement.

One of the biggest motivations for me to deprecate Query was that the union() case was so ambiguous and difficult to use.  usually when someone tries something "obvious" with query.union(), it just fails, in all kinds of bad ways, and even if it works, it builds an enormous nested query that is bad for both eyesight and database optimizers.  So switching to use select() where you can union() things together using union(), and allowing for better control of the entity you want to load, is what we did, and you can see the docs for this here: https://docs.sqlalchemy.org/en/14/orm/queryguide.html#selecting-entities-from-unions-and-other-set-operations 

the idea is that one part of your query expresses exactly what kind of object you want to load, in this case Car objects, and the other part is the SQL part, that is, the UNION of the two (or three) classes (add another class Limo and then do the same thing as Car/Truck with it):

```py
from sqlalchemy import select, union_all

stmt = select(Car).from_statement(
    union_all(
        select(Sedan),
        select(Truck),
        select(Limo)
    )
)

print(session.scalars(stmt).all())
```

that's how UNIONs look now.   the query you have certainly would have been nice to keep working in 1.4 as well, what you have there is a probably a regression of some kind.  however, i dont know if it's worth getting into, because it goes against how the class was supposed to work.

We can look at one great advantage of the new approach.   The ORM isn't going nuts trying to guess what you want, for the above, here's the SQL:

```sql
SELECT sedan.id, car.id AS id_1, car.type, sedan.wheel_count 
FROM car JOIN sedan ON car.id = sedan.id UNION ALL SELECT truck.id, car.id AS id_1, car.type, truck.wheel_count 
FROM car JOIN truck ON car.id = truck.id UNION ALL SELECT limo.id, car.id AS id_1, car.type, limo.wheel_count 
FROM car JOIN limo ON car.id = limo.id
```

perfect!  Whereas look at 1.3 with your other query (keep in mind this is three classes, not two, so is more complex than what you had):

```sql
SELECT anon_1.anon_2_sedan_id AS anon_1_anon_2_sedan_id, anon_1.anon_2_car_id AS anon_1_anon_2_car_id, anon_1.anon_2_car_type AS anon_1_anon_2_car_type, anon_1.anon_2_sedan_wheel_count AS anon_1_anon_2_sedan_wheel_count 
FROM (SELECT anon_2.sedan_id AS anon_2_sedan_id, anon_2.car_id AS anon_2_car_id, anon_2.car_type AS anon_2_car_type, anon_2.sedan_wheel_count AS anon_2_sedan_wheel_count 
FROM (SELECT sedan.id AS sedan_id, car.id AS car_id, car.type AS car_type, sedan.wheel_count AS sedan_wheel_count 
FROM car JOIN sedan ON car.id = sedan.id UNION SELECT truck.id AS truck_id, car.id AS car_id, car.type AS car_type, truck.wheel_count AS truck_wheel_count 
FROM car JOIN truck ON car.id = truck.id) AS anon_2 UNION SELECT limo.id AS limo_id, car.id AS car_id, car.type AS car_type, limo.wheel_count AS limo_wheel_count 
FROM car JOIN limo ON car.id = limo.id) AS anon_1
``` 

bleah!   awkward subqueries and all that, it looks pretty confused about the three classes and how to nest them, it's pretty awful.    A more equivalent query in 1.4 would be:

```sql
from sqlalchemy import select, union_all
from sqlalchemy.orm import aliased

car = aliased(
    Car, union_all(select(Sedan), select(Truck), select(Limo)).subquery()
)
stmt = select(car)

print(session.scalars(stmt).all())
```

but still vastly better

```sql
SELECT anon_1.id, anon_1.type 
FROM (SELECT sedan.id AS id, car.id AS id_1, car.type AS type, sedan.wheel_count AS wheel_count 
FROM car JOIN sedan ON car.id = sedan.id UNION ALL SELECT truck.id AS id, car.id AS id_1, car.type AS type, truck.wheel_count AS wheel_count 
FROM car JOIN truck ON car.id = truck.id UNION ALL SELECT limo.id AS id, car.id AS id_1, car.type AS type, limo.wheel_count AS wheel_count 
FROM car JOIN limo ON car.id = limo.id) AS anon_1
```

so...that's what I got.  explicit is better than implicit.   your query is *really* weird that it works though so maybe ill try to look some more to see why it returns the "correct" answer, but overall, that query shouldn't work and the error in 1.4 is kind of more what it "should" do.

**Reply by BobACollins** [reply]
Thanks for the detailed and quick response I managed to get our application moved over to 1.4! The sample code was very helpful! Now, we can start the slow transition to 2.0 style code. 🎉

**Comment by John-Sias** [comment]
This simple answer may not work for your probably more complex workflow, but what about:

`session.query(Car).all()`
or
`session.execute(select(Car)).scalars().all()`

These yield the two objects (Sedan and Truck)

or

`session.execute(union(select(Sedan),select(Truck))).all()`

yields the following rows:

```   
id  id_1   type  wheel_count
0   1     1  sedan            3
1   2     2  truck            4
```

Maybe, as I say, these only work because you simplified the example, but I offer them nonetheless.

John

**Reply by zzzeek** [reply]
sure, you can query for Car as well.  that's not "the same" as a UNION of two separate sibling classes, I didnt know if your real world example has 10 different subclasses, for example.
